### ISSUE TITLE
Flexible cable bugs

### ISSUE THREAD

**Issue by wormzjl** [Foghrye4/ihl#55]
https://pastebin.com/EVhRV2xS

**Comment by Foghrye4** [comment]
I can't recreate that crash.  Could you show me your setup?

**Comment by wormzjl** [comment]
It happens a bit random, probably by connecting ihl cable directly to ic2 cables

**Comment by Foghrye4** [comment]
I still could not catch it. I added additional null check in grid calculation in 0.640. This should solve a case.
http://www.mediafire.com/file/hgqj9ekcyagh457/ihl-tools_n_machines-0.640-1.7.10.jar

**Comment by wormzjl** [comment]
ok thx!

**Comment by yuyuyzl** [comment]
Another stall happened in our server today, luckily we have recorded the stall log and an image of his wire connections.
![ihlcablestall](https://cloud.githubusercontent.com/assets/14276008/26550580/f59e8dbe-44b0-11e7-94f7-10af8b5d3189.jpg)
And the stall log:
[stall-2017-05-29_20.44.19.txt](https://github.com/Foghrye4/ihl/files/1035831/stall-2017-05-29_20.44.19.txt)

Seems it would be a IHL&Uncomplication compatible problem.
And we're using Thermos.
Hoping this would help.

**Comment by Foghrye4** [comment]
Seems to be a fatal flaw of using NBT object as grid cable representative. Fixing this would take some time.

**Comment by Foghrye4** [comment]
Ok, try a new version (0.642), it should not cause stalls, because I get rid of NBT in energy net.

**Comment by tony987852** [comment]
It happened again. We have already updated to the latest version.

Here is the stall log:
[stall-2017-05-31_23.06.49-2.txt](https://github.com/Foghrye4/ihl/files/1042043/stall-2017-05-31_23.06.49-2.txt)

Thanks.

**Comment by Foghrye4** [comment]
I'm not sure in anything now, but I changed a code a little bit. Maybe this could help.
http://www.mediafire.com/file/5dx5ff93rmq152c/ihl-tools_n_machines-0.0.643-1.7.10.jar
BTW how complex is grid? Do it contain a circles?*
____________________________________
* that's should be absolutely safe.

**Comment by wormzjl** [comment]
It quiet simple, just power source -> electrolysis bath with length less than 2

**Comment by wormzjl** [comment]
https://github.com/Foghrye4/ihl/blob/8ed8c659da9496bf49fec94c049b0abae21746a8/ihl/flexible_cable/IHLGrid.java#L272 
Can you add some sort of size limit & debug warning on this hashmap? Might help pinpointing the issue
Also is it possible to print the location of bugged cable?

**Comment by Foghrye4** [comment]
@wormzjl I removed that hashmap. Source updated, check by yourself.

**Comment by wormzjl** [comment]
ok will test

**Comment by wormzjl** [comment]
Seems that 0.643 still has the old code?

**Comment by Foghrye4** [comment]
@wormzjl ...no. Does it stall?

**Comment by wormzjl** [comment]
Nvm, still testing
Just have a probable cause for the issue coming up in mind.
Thermos/ spigot "deactivates" entities if they are some distance away from players, will this cause issue with the cables?
Also if the nbt is not saved properly on chunk unload, on chunk reload will it crash like this?

**Comment by Foghrye4** [comment]
@wormzjl entities does not involved in enegy transfer unless they are touch ground and uninsulated or voltage in cable above insulation breakdown voltage. In that case they emit smoke particles.
Each time chunk containing energy node unloaded ... wait... I think I found a possible case. When TE is in uloaded chunk it is invalidated and removed from all lists. But in my case it is remains in grid TE list. So, when chunk is loaded again it will generate new instance of TE, which will be added to grid alongside with old invalid instance of TE.

**Comment by wormzjl** [comment]
So boom?

**Comment by Foghrye4** [comment]
Ok, try this: http://www.mediafire.com/file/taojc7ubcicbyke/ihl-tools_n_machines-0.644-1.7.10.jar
Also check fractionator column.

**Comment by wormzjl** [comment]
https://pastebin.com/PSijj7ht
crashed by linking an electrolysis bath with a mfe

**Comment by Foghrye4** [comment]
Unexpected.
Well this version does not crash: http://www.mediafire.com/file/gw2fbm0yl7j7o5y/ihl-tools_n_machines-0.645-1.7.10.jar
(and as well does not work with uncomplication mod (energy not transferred after first packet))

**Comment by wormzjl** [comment]
Some weirdness in the enet?

**Comment by Foghrye4** [comment]
@wormzjl probably demanded energy >0d but <1d bug.

**Comment by wormzjl** [comment]
Seems that when the electrolysis bath finishes it's work, it refuses to work again unless the cable is reconnected

**Comment by Foghrye4** [comment]
For a freaking hour I tried one to another options to fix that. Screw it. Ok, IHL officially incompatible with Uncomplication mod until a further notice.

Maybe I should make electrolyzer to support work with normal cables?

**Comment by wormzjl** [comment]
That will lose some...identity

**Comment by wormzjl** [comment]
Anyway the crash seems to be fixed, no more crashes today

**Comment by Foghrye4** [comment]
@wormzjl I took my will into fist and found a lot of bugs related to flexible cable energy transfer system (0.646). I did not test fixes with uncomplication, but it was not an uncomplication related bugs at a first place. You may want to check a work with uncomplication now, and I need to sleep.

**Comment by wormzjl** [comment]
Thank you very much! gonna do some more testing

**Comment by wormzjl** [comment]
Ok, the power transfer from/to ic2 energy block is fine.
But it seems that the electrolysis bath is not receiving any power?
Also before you can attach wire to desired block face by simple shift + rightclick, but now you need to stand right before that side

**Comment by Foghrye4** [comment]
>But it seems that the electrolysis bath is not receiving any power?

Is recipe conditions are fulfilled? Because everything fine at my side (even with uncomplication mod). Try this: launch Minecraft with opened console. Right-click at electrolysis bath with thermometer (from IHL creative inventory). Copy-paste 4 lower lines of console output here. Do a same thing with opposite side of cable connected (there should be invisible but clickable "junction box").

>Also before you can attach wire to desired block face by simple shift + rightclick, but now you need to stand right before that side

O.O I did not change anything related to this part of code. At least as I remember.

**Comment by yuyuyzl** [comment]
> Also before you can attach wire to desired block face by simple shift + rightclick, but now you need to stand right before that side  

That's true. I have not read your code but in game you should stand at the face you want to put wire and click or your wire will connect to other faces.

> Because everything fine at my side (even with uncomplication mod).

And i'm sorry about my misleading to uncomplication. IC+IHL+Uncomplication works fine, The problem is Immersive engineering use a strange way to feed energy to IC capacitors, which causes the problem i've mentioned before. An IE->IC->IHL wire connection will cause the problem, but IE->IC->IC->IHL wire works fine.

**Comment by wormzjl** [comment]
[19:34:03] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:104]: gridId=7
[19:34:03] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:106]: grid energy=0.0
[19:34:03] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:107]: getEnergyAmountThisNodeWant()=32000.0
[19:34:04] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:104]: gridId=7
[19:34:04] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:106]: grid energy=0.0
[19:34:04] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:107]: getEnergyAmountThisNodeWant()=32000.0
[19:34:05] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:104]: gridId=7
[19:34:05] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:106]: grid energy=0.0
[19:34:05] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:107]: getEnergyAmountThisNodeWant()=32000.0

The recipe works normally before update

setup
![image](https://user-images.githubusercontent.com/2989009/27096180-7634ad30-5070-11e7-9c6d-d50c994af0b6.png)

**Comment by Foghrye4** [comment]
>getEnergyAmountThisNodeWant()=32000.0

32000 at both sides?
This value show amount of energy which can consume node when energy injected in net.
400*400/5 - is basic value before grid update for electrolisis bath. So everything fine here.
Please check opposite side (invisible junction box). There will be a small black box when you point tool to an end of cable.
http://imgur.com/a/bgJk1

**Comment by wormzjl** [comment]
[20:09:51] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:88]: demanded=0.0
[20:09:51] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:104]: gridId=1
[20:09:51] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:106]: grid energy=0.0
[20:09:51] [Server thread/INFO] [STDOUT/]: [ihl.items_blocks.IHLTool:func_77659_a:107]: getEnergyAmountThisNodeWant()=0.0

This is on the side of the MFE

![image](https://user-images.githubusercontent.com/2989009/27097381-59ce057a-5074-11e7-86bf-18fb0c30aae8.png)

and the wire I'm using
![image](https://user-images.githubusercontent.com/2989009/27097556-d9f95fd8-5074-11e7-9a65-1338e60aee49.png)

**Comment by Foghrye4** [comment]
Weird. Different grid ids... (id=1 at MFE side and id=7 at electrolysis bath side). Is it bug persistent? If you remove a cable and reconnect devices together again it will be a same?

**Comment by wormzjl** [comment]
it's the same all the time

**Comment by Foghrye4** [comment]
@wormzjl send me your world and modpack.

**Comment by wormzjl** [comment]
https://mega.nz/#!fcRGRCwC!MA62EeTdt_yY18IVzBUyDY6AVMF4sedyKcYnVOZzztY
You can import it via multimc

**Comment by Foghrye4** [comment]
@wormzjl it's a DragonAPI. Have no idea what it do, but everything works fine without it.

**Comment by wormzjl** [comment]
why it no longer works after the fix...this is really weird-

**Comment by Foghrye4** [comment]
@wormzjl A thing I fixed - "junction box" sending energy back to source if it is a sink as well (all energy storage devices are both sink and sources). If it is a second node in grid - everything work fine. If it is first - energy does not go further. I added condition to junction box to send energy to attached device only if this device allow input from box side.

**Comment by wormzjl** [comment]
Seems that the cable orientation issue starts from 645...still checking

**Comment by wormzjl** [comment]
If the machines and wire were placed before adding dragonapi, everything works properly...investigating

**Comment by wormzjl** [comment]
Nbt for the one that is working:
![image](https://user-images.githubusercontent.com/2989009/27240093-c4f4ac12-52d3-11e7-88ee-b7f61adc7e05.png)

Nbt for the one that is NOT working:
![image](https://user-images.githubusercontent.com/2989009/27240135-e6e46f42-52d3-11e7-8ad4-ce052d8a5cf2.png)

**Comment by wormzjl** [comment]
OK, if I remove the voltz api from the dragonapi jar everything works properly
After looking at the code...method name duplication -.-? Is this even possible ?
![image](https://user-images.githubusercontent.com/2989009/27240370-b83e8da2-52d4-11e7-97a5-fdf6c82845f6.png)

**Comment by Foghrye4** [comment]
Damn. It seems you founded it. If DragonAPI use ASM to add that interface with provided functions into all IC2 energy tiles, this could cause collision with my own function result. And it seems DragonAPI just catch exceptions caused by this mess silently.

**Comment by wormzjl** [comment]
No this is probably not the issue so I deleted it, still investigating

**Comment by wormzjl** [comment]
Ok this has nothing to do with dragonapi
Ihl cables only conducts power when energy source is facing east, you might want to have a look at the side checks

**Comment by Foghrye4** [comment]
You mean that energy source transfer energy to flexible cables only if it facing east? Why everything works fine without DragonAPI then?

**Comment by wormzjl** [comment]
yeah, tested without DragonAPI, no power emit when it's not facing east even with MFE->MFE connection, nothing to do with the electrolysis bath

**Comment by Foghrye4** [comment]
Ooops. My bad.

**Comment by Foghrye4** [comment]
What about now? 
http://www.mediafire.com/file/zq5pffyh311hibo/ihl-tools_n_machines-0.647-1.7.10.jar

**Comment by wormzjl** [comment]
thx, doing some live testing

**Comment by wormzjl** [comment]
no problem occurred yet, seems to be solved

**Comment by wormzjl** [comment]
Crashing again...gonna post them later

**Comment by tony987852** [comment]
New bug:
When cable overload, it will crash the server.
I also wonder why it did not happen before today that using the same cable and the same energy network.

I tried 3 times.
[crash-2017-07-06_10.02.03-server.txt](https://github.com/Foghrye4/ihl/files/1127350/crash-2017-07-06_10.02.03-server.txt)
[crash-2017-07-06_10.20.48-server.txt](https://github.com/Foghrye4/ihl/files/1127349/crash-2017-07-06_10.20.48-server.txt)
[crash-2017-07-06_10.31.45-server.txt](https://github.com/Foghrye4/ihl/files/1127351/crash-2017-07-06_10.31.45-server.txt)

Thanks.

**Comment by tony987852** [comment]
The stall happened again. I don't know which player is using the cable.

[stall-2017-07-06_21.18.46.txt](https://github.com/Foghrye4/ihl/files/1128025/stall-2017-07-06_21.18.46.txt)
[stall-2017-07-06_19.39.39.txt](https://github.com/Foghrye4/ihl/files/1128024/stall-2017-07-06_19.39.39.txt)

**Comment by Foghrye4** [comment]
I wonder how that is happened. I add a code to prevent crash in 0.650 and it will generate some log output while removing corrupted tile-entities. I would like to see that log. As for stalls - if they happened again I would like to see log output as well.

**Comment by wormzjl** [comment]
[crash-2017-07-08_16.06.36-server.txt](https://github.com/Foghrye4/ihl/files/1132844/crash-2017-07-08_16.06.36-server.txt)
[fml-server-3.txt](https://github.com/Foghrye4/ihl/files/1132846/fml-server-3.txt)
Here is crashlog with fml log, seems to be caused by cable burning out

Also another bug, seems that the length of cable used is calculated by player move distance?
If I do not move during cable placement, the length will be 0

**Comment by Foghrye4** [comment]
@wormzjl is latest version of IHL fix issue?

**Comment by wormzjl** [comment]
Been able to crash it on singleplayer, put down a MFE, connect it to a MV transformer then a diamond cable, then ihl cable -> electrolysis bath. Crash seems to happen on cable meltdown.
[crash-2017-07-08_16.22.51-server.txt](https://github.com/Foghrye4/ihl/files/1132853/crash-2017-07-08_16.22.51-server.txt)
[fml-client-3.txt](https://github.com/Foghrye4/ihl/files/1132852/fml-client-3.txt)



Also got another bug, seems that the length of cable used is calculated by player move distance?
If I do not move during cable placement, the length will be 0

**Comment by wormzjl** [comment]
ihl-tools_n_machines-0.648-1.7.10.jar
Umm you have updated to 650? did anything change related to the cable?
Going to test with the latest

**Comment by Foghrye4** [comment]
@wormzjl yes. I added a code, which should remove nodes with ID == -1 from grid and generate some log output.

**Comment by wormzjl** [comment]
Ok, can you take a look at the cable length issue mentioned above?

**Comment by wormzjl** [comment]
Got the same result.....
[fml-server-3.txt](https://github.com/Foghrye4/ihl/files/1132861/fml-server-3.txt)

**Comment by Foghrye4** [comment]
Damn, I forgot to add a one more line.

**Comment by Foghrye4** [comment]
http://www.mediafire.com/file/6565zoj5and4sna/ihl-tools_n_machines-0.651-1.7.10.jar
Ok, it should work now.

**Comment by wormzjl** [comment]
[17:05:27] [Server thread/ERROR] [ihl/]: Removing TE with grid ID==-1 ihl.processing.chemistry.ElectrolysisBathTileEntity If it happened more than a single time, report that to mod author.
[17:05:27] [Server thread/WARN] [Uncomplication/]: EnergyNet.emitEnergyFrom: ihl.flexible_cable.AnchorTileEntity@2d2179 is not added to the enet
[17:05:40] [Server thread/ERROR] [ihl/]: Removing TE with grid ID==-1 ihl.processing.chemistry.ElectrolysisBathTileEntity If it happened more than a single time, report that to mod author.
[17:05:40] [Server thread/WARN] [Uncomplication/]: EnergyNet.emitEnergyFrom: ihl.flexible_cable.AnchorTileEntity@2d2179 is not added to the enet
[17:05:53] [Server thread/ERROR] [ihl/]: Removing TE with grid ID==-1 ihl.processing.chemistry.ElectrolysisBathTileEntity If it happened more than a single time, report that to mod author.
[17:05:53] [Server thread/WARN] [Uncomplication/]: EnergyNet.emitEnergyFrom: ihl.flexible_cable.AnchorTileEntity@2d2179 is not added to the enet
[17:05:59] [Server thread/ERROR] [ihl/]: Removing TE with grid ID==-1 ihl.processing.chemistry.ElectrolysisBathTileEntity If it happened more than a single time, report that to mod author.
[17:05:59] [Server thread/WARN] [Uncomplication/]: EnergyNet.emitEnergyFrom: ihl.flexible_cable.AnchorTileEntity@2d2179 is not added to the enet

Got these in the console 

Screenshot, the cable can be actually burned
![1](https://user-images.githubusercontent.com/2989009/27984133-d9a845f0-63cd-11e7-9ea8-894a29e5fdbc.jpg)

![2](https://user-images.githubusercontent.com/2989009/27984135-dccfc1ea-63cd-11e7-8bf3-332ca2b2a9aa.jpg)

**Comment by Foghrye4** [comment]
So happen every time on cable burnout?

**Comment by wormzjl** [comment]
Nope, seems to happen only on certain setups

**Comment by wormzjl** [comment]
![image](https://user-images.githubusercontent.com/2989009/27984580-bbb992a4-63d9-11e7-82a7-ca14d0915f92.png)

Seems somewhat related to direction, the one on the left will not give error on burnout, the one on the right will

**Comment by wormzjl** [comment]
![image](https://user-images.githubusercontent.com/2989009/27984593-225e6890-63da-11e7-848b-ccfaa4875d73.png)
Also...the cable length bug

**Comment by wormzjl** [comment]
This is too difficult to replicate, maybe just let the electrolysis bath accept EU directly (from the upper side)

**Comment by Foghrye4** [comment]
http://www.mediafire.com/file/5ojxiv8p9h3tgk3/ihl-tools_n_machines-0.652-1.7.10.jar
